### PR TITLE
Make Bastion controller and BackupEntry component ready for cached client

### PR DESCRIPTION
**Pull request by timebertt**
**How to categorize this PR?**

/area dev-productivity scalability
/kind cleanup

**What this PR does / why we need it**:

Similar to previous PRs, this PR works towards getting rid of the `DirectClient`.
This one tackles usages in the `Bastion` controller and `core.BackupEntry` component and makes them ready for cached clients.

**Which issue(s) this PR fixes**:
Part of #2822

**Special notes for your reviewer**:

/squash

**Release note**:

```other operator

```

### CODE REVIEW

**Comment by timebertt** [comment]
Rebased to solve conflicts

**Comment by xrstf** [comment]
Maybe this is the right PR to fix another "bug" I encountered.

I start my local gardener stack (gapi, g-ctrlmgr, g-scheduller, g-gardenlet) and let them warm up for a minute. Then I create a new `Bastion` resource on the garden and noticed that the gardenlet is not syncing it to the seed. I have to restart the gardenlet to make it work.

_Deleting_ the bastion however is picked up immediately by the gardenlet. Only creations and maybe updates do not trigger a reconciliation at this time. This happens both with master and your branch.

File: pkg/extensions/customresources.go [hunk 1]
```diff
@@ -20,25 +20,25 @@ import (
 	"fmt"
 	"time"
 
+	"github.com/sirupsen/logrus"
+	autoscalingv1 "k8s.io/api/autoscaling/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
```
Comment (xrstf) [inline pkg/extensions/customresources.go:29]: Did the project switch the general guideline for import orders? It was more or less "std, project, external" so far and this turns it into "std, external, project".

**Comment by timebertt** [comment]
> Maybe this is the right PR to fix another "bug" I encountered.

Hmm, I'm not sure how this is related. Do you have an idea, what causes this?
If not, I would prefer to not block this PR because of this, as you said it also happens on master.

File: pkg/extensions/customresources.go (hunk 1 above)
Comment (timebertt) [reply pkg/extensions/customresources.go:29]: Well, the guidelines we agreed on was to accept the import order as long as `goimports` doesn't complain.
Sometimes my IDE put stuff in reverse order, sometimes it does not. I don't know why, but I also don't care to much, as both is accepted by `goimports`...

**Comment by xrstf** [comment]
> If not, I would prefer to not block this PR because of this, as you said it also happens on master.

Yeah this should not block, I just speculated that I didn't set up some informer right and this PR might be related. Now that I read it, it isn't really.

Except for the changed import order this looks mostly mechanical here. ^^

File: pkg/extensions/customresources.go (hunk 1 above)
Comment (xrstf) [reply pkg/extensions/customresources.go:29]: Okay then @rfranzke should not have told me to re-order imports at all :-D I was under the impression that there is a somewhat agreed upon order in gardener/* repos.

**Comment by timebertt** [comment]
Rebased onto current master.

File: pkg/operation/botanist/component/backupentry/backupentry.go [hunk 2]
```diff
@@ -159,36 +149,25 @@ func (b *backupEntry) WaitMigrate(ctx context.Context) error {
...
 		metav1.SetMetaDataAnnotation(&backupEntry.ObjectMeta, v1beta1constants.GardenerOperation, operation)
 		metav1.SetMetaDataAnnotation(&backupEntry.ObjectMeta, v1beta1constants.GardenerTimestamp, TimeNow().UTC().String())
+
 		if b.values.ShootPurpose != nil {
 			metav1.SetMetaDataAnnotation(&backupEntry.ObjectMeta, v1beta1constants.ShootPurpose, string(*b.values.ShootPurpose))
 		}
 
-		finalizers := sets.NewString(backupEntry.GetFinalizers()...)
-		finalizers.Insert(gardencorev1beta1.GardenerName)
-		backupEntry.SetFinalizers(finalizers.UnsortedList())
```
Comment (rfranzke) [inline pkg/operation/botanist/component/backupentry/backupentry.go:190]: Why was this removed?

Comment (timebertt) [reply pkg/operation/botanist/component/backupentry/backupentry.go:190]: The backupentry controller adds its finalizer to the object [here](https://github.com/gardener/gardener/blob/b721f7f118f2d4cdb33ee0020f213a9fab6078eb/pkg/gardenlet/controller/backupentry/backup_entry_reconciler.go#L101), so there shouldn't be the need to add it here when creating the `BackupEntry` object. Also, this is semantically incorrect, as the finalizer should be added by the controller and not the applier, right?

Comment (rfranzke) [reply pkg/operation/botanist/component/backupentry/backupentry.go:190]: Ah, sure, yes, thanks for the explanation and for cleaning up!

**Review by rfranzke — APPROVED** [review]
/lgtm
